proposals/file-types.md

```diff
@@ -0,0 +1,188 @@
+# File types
+
+## Summary
+[summary]: #summary
+
+Permit a `file` modifier on top-level type declarations. The type only exists in the file where it is declared.
+
+```cs
+// File1.cs
+namespace NS;
+
+file class Widget
+{
+}
+
+// File2.cs
+namespace NS;
+
+file class Widget // different symbol than the Widget in File1
+{
+}
+
+// File3.cs
+using NS;
+
+var widget = new Widget(); // error: The type or namespace name 'Widget' could not be found.
+```
+
+## Motivation
+[motivation]: #motivation
+
+Our primary motivation is from source generators. Source generators work by adding files to the user's compilation.
+1. Those files should be able to contain implementation details which are hidden from the rest of the compilation, yet are usable throughout the file they are declared in.
+2. We want to reduce the need for generators to "search" for type names which won't collide with declarations in user code or code from other generators.
+
+## Detailed design
+[design]: #detailed-design
+
+- We add the `file` modifier to the [class](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/classes.md#1422-class-modifiers), [struct](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/structs.md#1522-struct-modifiers), [interface](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/interfaces.md#1722-interface-modifiers), and [enum](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/enums.md#183-enum-modifiers) modifier sets in the grammar.
```

Comment:
And `record`?

Reply:
Yeah, I think record, record struct and delegate should be included here.